treeherder/auth/backends.py

```diff
@@ -0,0 +1,148 @@
+                    'secrets:set:project/treeherder/*'],
+         'scheme': 'hawk',
+         'clientId': 'mozilla-ldap/biped@mozilla.com',
+         'expires': '2016-10-31T17:40:45.692Z'}
+    """
+
+    def _get_scope_value(self, result, scope_prefix):
+        for scope in result["scopes"]:
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        """
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+
+        if not email or email == "*":
+            # try finding the email in the clientId.  It can then be used to
+            # find a matching user.
+            match = re.search(
+                r"([a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)",
+                result["clientId"])
+            if match:
+                email = match.group(0)
+            else:
+                # if we STILL don't have an email at this point, then just
+                # use their clientId as a dummy email.  We don't
+                # send anything to it, so it doesn't matter much, in
+                # reality.  This can only happen if the user isn't using LDAP,
+                # and doesn't have an email in their clientId.
+                email = result["clientId"]
+        return email
+
+    def _get_user(self, username, email):
+        """
+        Try to find an exising user that matches either the username
+        or email.  Prefer the username, since that's the unique key.  But
+        fallback to the email.
+
+        """
+
+        # Since there is a unique index on username, but not on email,
+        # it is POSSIBLE there could be two users with the same email and
+        # different usernames.  Not very likely, but this is safer.
+        user = User.objects.filter(
+            Q(username=username) | Q(email=email))
+
+        # if we didn't find either, then raise an exception so we create a new
+        # user
+        if not len(user):
+            raise ObjectDoesNotExist
+
+        # prefer a matching username.  If no username match, return the first
+        # user in the list.
+        try:
+            return user.get(username=username)
+        except ObjectDoesNotExist:
+            return user.first()
+
+    def authenticate(self, authorization=None, host=None, port=None):
+        user = None
+
+        if authorization:
+            tc_auth = Auth()
+            result = tc_auth.authenticateHawk({
+                "authorization": authorization,
+                "host": host,
+                "port": port,
+                "resource": "/",
+                "method": "get",
+            })
+
+            if result["status"] == "auth-success":
+                # TODO: remove this size limit when we upgrade to django 1.10
+                # in Bug 1311967
+                username = result["clientId"][-30:]
```

Comment:
Perhaps we should continue to use a hash or similar, to prevent someone's truncated email matching? (Or just blocking long email addresses - Gerard's long email was a gmail, I suspect he won't be able to use that now, since his Okta with be @mozilla.com)

Speaking of which, I guess we'll need to manually migrate people whose LDAP is not the same as their Persona email (only if they need specific sheriff/admin permissions or are credential owners). Happy to just do that post-deploy.